lib/bolt/transport/ssh/exec_connection.rb

```diff
@@ -0,0 +1,123 @@
+        end
+
+        def build_ssh_command(command)
+          ssh_conf = @target.transport_config['ssh-command']
+          ssh_cmd = if ssh_conf.is_a?(String)
+                      [ssh_conf]
+                    elsif ssh_conf.is_a?(Array)
+                      ssh_conf
+                    end
+          ssh_cmd += %W[-p #{@target.port}] if @target.port
+          ssh_cmd += %w[-o BatchMode=yes]
+          ssh_cmd += %w[-o StrictHostKeyChecking=no] if @target.transport_config['host-key-check'] == false
+          if (key = target.transport_config['private-key'])
+            if key.instance_of?(String)
+              ssh_cmd += %W[-i #{key}]
+            else
+              # This is simpler than writing out a tempfile for now
+              raise Bolt::ValidationError, 'private-key must be a filepath when using ssh-command'
+            end
+          end
+          ssh_cmd << userhost
+          ssh_cmd << command
+          Shellwords.shelljoin(ssh_cmd)
+        end
+
+        def copy_file(source, dest)
+          @logger.debug { "Uploading #{source}, to #{userhost}:#{dest}" } unless source.is_a?(StringIO)
+          cp_cmd = @target.transport_config['copy-command'] || ["scp", "-r"]
+          cp_cmd += %W[-P #{@target.port}] if @target.port
+          cp_cmd += %w[-o StrictHostKeyChecking=no] if @target.transport_config['host-key-check'] == false
+          if (key = target.transport_config['private-key'])
+            if key.instance_of?(String)
+              cp_cmd += %W[-i #{key}]
+            else
+              # This is simpler than writing out a tempfile for now
+              raise Bolt::ValidationError, 'private-key must be a filepath when using ssh-command'
+            end
+          end
+
+          _, err, stat = if source.is_a?(StringIO)
+                           Tempfile.create(File.basename(dest)) do |f|
+                             f.write(source.read)
+                             f.close
+                             cp_cmd << f
+                             cp_cmd << "#{userhost}:\"#{dest}\""
+                             Open3.capture3(Shellwords.shelljoin(cp_cmd))
+                           end
+                         else
+                           cp_cmd << source
+                           cp_cmd << "#{userhost}:\"#{dest}\""
+                           Open3.capture3(Shellwords.shelljoin(cp_cmd))
+                         end
+
+          if stat.success?
+            @logger.debug "Successfully uploaded #{source} to #{dest}"
+          else
+            message = "Could not copy file to #{dest}: #{err}"
+            raise Bolt::Node::FileError.new(message, 'COPY_ERROR')
+          end
+        end
+
+        def execute(command)
+          Open3.popen3(build_ssh_command(command))
+        end
+
+        # This is used by the Bash shell to decide whether to `cd` before
+        # executing commands as a run-as user
+        def reset_cwd?
+          false
+        end
+
+        def validate(target)
```

Comment:
Is there a reason not to have this in `Bolt::Config::Transport::SSH`? The first part could be gated on whether `ssh-command` is set and the second part looks like it should apply to both connection types.